app/services/user_service.py

```diff
@@ -243,6 +245,190 @@ async def get_user_preferences(self, user_id: str) -> Dict[str, Any]:
+                        "apikey": self.api_key,
+                        "Authorization": f"Bearer {self.api_key}",
+                        "Content-Type": "application/json"
+                    },
+                    params={
+                        "id": f"eq.{user_id}"
+                    }
+                )
+
+                if response.status_code not in (200, 201, 204):
+                    error_detail = "Failed to retrieve user profile"
+                    try:
+                        error_data = response.json()
+                        if "message" in error_data:
+                            error_detail = error_data["message"]
+                    except Exception:
+                        pass
+
+                    logger.error(f"Profile retrieval failed: {error_detail}")
+                    raise HTTPException(
+                        status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
+                        detail=f"Failed to retrieve user profile: {error_detail}"
+                    )
+
+                profile_data = response.json()[0]
+                return UserProfile(**profile_data)
+
+        except httpx.RequestError as e:
+            logger.error(f"Request error retrieving profile: {str(e)}")
+            raise HTTPException(
+                status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
+                detail=f"Error communicating with database: {str(e)}"
+            )
+        except Exception as e:
+            logger.error(f"Unexpected error retrieving profile: {str(e)}")
+            raise HTTPException(
+                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
+                detail=f"Error retrieving user profile: {str(e)}"
+            )
+        
+    async def update_user_profile(self, user_id: str, user_data: UpdateUserProfileRequest) -> UserProfile:
+        """Update user profile.
+
+        Args:
+            user_id: Supabase user ID
+            user_data: User data to be updated
+
+        Returns:
+            User preferences dictionary
+
+        Raises:
+            HTTPException: If there is an error updating user profile
+        """
+        logger.info(f"Updating profile for user: {user_id}")
+
+        try:
+            user_profile = remove_null_values(user_data.model_dump())
+            async with httpx.AsyncClient() as client:
+                response = await client.patch(
+                    f"{self.base_url}/rest/v1/user_profiles",
+                    headers={
+                        "apikey": self.api_key,
+                        "Authorization": f"Bearer {self.api_key}",
+                        "Content-Type": "application/json",
+                        "Prefer": "return=representation"
+                    },
+                    params={
+                        "id": f"eq.{user_id}"
+                    },
+                    json = user_profile
+                )
+
+                if response.status_code not in (200, 201, 204):
```

Comment:
I think we should only check status 200 and 204. 201 could be valid if we are updating or creating if not exist